core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1133,10 +1182,71 @@ public void listStatus(AlluxioURI path, ListStatusContext context,
   public List<FileInfo> listStatus(AlluxioURI path, ListStatusContext context)
       throws AccessControlException, FileDoesNotExistException, InvalidPathException, IOException {
     final List<FileInfo> fileInfos = new ArrayList<>();
-    listStatus(path, context, (item) -> fileInfos.add(item));
+    listStatus(path, context, fileInfos::add);
```

Comment:
I'm feeling this `listStatus` is getting more clumsy cuz we keep adding new logic and if-conditions. Will it be possible to make the split here like 
```
public List<FileInfo> listStatus(AlluxioURI path, ListStatusContext context)
      throws AccessControlException, FileDoesNotExistException, InvalidPathException, IOException {
  if (partialListing) {
    listStatusPartial();
  } else {
    listStatusFull();
  }
```

The new `listStatusPartial()` and `listStatusFull()` will have a lot of overlap in code. That's fine because for each overlap we can extract a common part into a method. So after the re-org I imagine `listStatusPartial()` will be like:
```
listStatusPartial() {
  // some specific logic
  ...
  if (syncNeeded) {
    sync();
  }

  doSomethingCommon();

  // Some more specific logic
  ...

  doSomethingMoreCommon();
}
```

Reply:
I tried to clean it up a bit by moving the partial listing logic into their own functions.